New __iter__  for Array module

kangzhiq opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed
The array iteration is modified to have a new behavior:
Before:
```python
>>> a = Array([[1, 2], [3, 4]])
>>> for i in a:
...     print(i)
...
1
2
3
4
 
```
Now:
```python
>>> a = Array([[1, 2], [3, 4]])
>>> for i in a:
...     print(i)
...
[1, 2]
[3, 4]
```
Besides, the iteration over each item is moved to a new object named `nditer` :
```python
>>> from sympy.tensor.array.nditer import nditer
>>> for i in nditer(a):
...     print(i)
...
1
2
3
2
3
4
```

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* tensor
  * Modified __iter__ function and added new class named nditer
<!-- END RELEASE NOTES -->

Commit:
New Iteration for Array

- Modified iteration function inside Array
- Added nditer object to iterate every element
- Modified _getitem_ to fix the case of A[:]
- Modified code and tests accordingly

Commit:
Remove changes for iterating list

kangzhiq commented on sympy/tensor/array/nditer.py:
```diff
@@ -0,0 +1,53 @@
+
+        try:
+            if isinstance(self._iter, list):
+                result = self._iter[self._idx]
```
How to dynamically handle the iteration over a multi-dimension list is a great challenge. Especially for the case where an `array` is cast to `list`:
```python
>>> a = Array([[1, 2], [3, 4]])
>>> b = list(a)
>>> b.__class__
<class 'list'>
>>> b[0].__class__
<class 'sympy.tensor.array.dense_ndim_array.ImmutableDenseNDimArray'>
```
I might need some guidance in this problem. :-)

Upabjojr replied:
Please ignore `list`. We have `Array` in SymPy. People who use `list` will use 99% of the time `NumPy`, we don't need support for `list`.

Upabjojr replied:
We don't need to redo what NumPy already does... if you want to use NumPy's features, just use NumPy.

kangzhiq replied:
But we have tests like:
```
 assert list(sparse_array) == [0, 0, 0, 1]
```
That is why I ran into an error about `list` and tried to handle this special case. However, I have changed the test by using a simple `tolist()` instead of 'list()'.

kangzhiq commented on sympy/tensor/tests/test_tensor.py:
```diff
@@ -1318,19 +1318,19 @@ def test_valued_tensor_iter():
-    assert list(-5 * BA(i1, i2)) == [-5 * i for i in list(ba_matrix)]
+    #assert list(BA(i1, i2)) == list(ba_matrix)
+    #assert list(3 * BA(i1, i2)) == [3 * i for i in list(ba_matrix)]
+    #assert list(-5 * BA(i1, i2)) == [-5 * i for i in list(ba_matrix)]
```
New tests might be needed.

Commit:
Fixed bug of list(Array())

Upabjojr commented on sympy/tensor/array/nditer.py:
```diff
@@ -0,0 +1,58 @@
+from sympy.tensor.array.ndim_array import NDimArray
+from sympy.core.compatibility import Iterable
+
+class nditer(object):
```
SymPy standard is `NDIter`

Upabjojr replied:
This should be a SymPy object, so inherit `Expr`.

Upabjojr replied:
Maybe this object should be called `Flatten`, as what it does it flattening the array.

Upabjojr commented on sympy/tensor/array/nditer.py:
```diff
@@ -0,0 +1,58 @@
+
+        if not isinstance(iterable, (Iterable, MatrixBase)):
+            raise NotImplementedError("Data type not yet supported")
+        if isinstance(iterable, list) and isinstance(iterable[0], NDimArray):
```
why this condition?

kangzhiq replied:
This condition is to handle to case where an array is cast to list by `list()`:
```python
>>> a = list(Array([[0, 1], [2, 3]]))
>>> a.__class__
<class 'list'>
>>> a[0].__class__
<class 'sympy.tensor.array.dense_ndim_array.ImmutableDenseNDimArray'>
```
And the multdimensional `list` is not considered here, as you explained below. I was thinking the same so I only handled this special case of having a mutipledimensional `list` from `list()`.

Upabjojr commented on sympy/tensor/array/nditer.py:
```diff
@@ -0,0 +1,58 @@
+from sympy.tensor.array.dense_ndim_array import DenseNDimArray
```
please no new files, there is a file for array operations.

Upabjojr commented on sympy/tensor/array/nditer.py:
```diff
@@ -0,0 +1,58 @@
+        self._idx = 0
+
+    def __iter__(self):
+        return self
```
why not return the iterator here?

kangzhiq replied:
I was thinking the `NDIter` itself as an iterator, that is why `self` is returned here.

Upabjojr commented on sympy/tensor/array/nditer.py:
```diff
@@ -0,0 +1,58 @@
+    def __next__(self):
+        from sympy.matrices.matrices import MatrixBase
+
+        try:
```
No `try` please.

kangzhiq replied:
I was using `try` to handle the `IndexError` so that the iterator can be stopped properly. I have seen the usage of `try+except` structure in SymPy's codebase. Should it be avoided?

Commit:
Renamed class to Flatten

- Renamed class to Flatten
- Fixed a bug of index

Upabjojr commented on sympy/tensor/array/arrayop.py:
```diff
@@ -296,3 +296,57 @@ def permutedims(expr, perm):
     return type(expr)(new_array, new_shape)
+
+
+class Flatten(object):
```
inherit `Basic`

Upabjojr commented on sympy/tensor/array/arrayop.py:
```diff
@@ -296,3 +296,57 @@ def permutedims(expr, perm):
+
+        if not isinstance(iterable, (Iterable, MatrixBase)):
+            raise NotImplementedError("Data type not yet supported")
+        if isinstance(iterable, list) and len(iterable)>0 and isinstance(iterable[0], NDimArray):
```
if it's a list, then transform it into an `Array`.

kangzhiq replied:
If we transform it to `Array`, there will be a loop of calling functions:
While initiating an Array, we will call `Flatten()` in return.

Upabjojr replied:
You cannot add a `list` to a SymPy object. What about calling `flatten( )` in `Array`?

kangzhiq replied:
OK!

Commit:
Test passed

Fixed error in diff()

Commit:
Modified heritage to Basic

Upabjojr commented on sympy/tensor/array/dense_ndim_array.py:
```diff
@@ -64,13 +64,13 @@ def __getitem__(self, index):
-                return self._array[index]
+            # if isinstance(index, slice):
+                # return self._array[index]
+            # else:
```
if you are removing code, just remove it

Upabjojr commented on sympy/tensor/array/dense_ndim_array.py:
```diff
@@ -148,11 +155,12 @@ def __new__(cls, iterable, shape=None, **kwargs):
         shape = Tuple(*map(_sympify, shape))
         cls._check_special_bounds(flat_list, shape)
-        flat_list = flatten(flat_list)
+        flat_list = flatten(Flatten(flat_list))
```
is this really needed? isn't `flatten` already supposed to flatten the list?

kangzhiq replied:
No, it can be removed. I didn't think about it. Thank you!!
Since I have modified the `flatten()` function in the last commit, it can now flatten the result of `list(Array())`. 
What we had before is that it would run into the case of having 'args' and transform an `Array` to `Arrays.args`, which will add the `shape` into the flatten list.
```python
 if hasattr(el, 'args'):
    el = el.args
```

Commit:
Updated code and added doctest and tests

- Updated code for simplification
- Added documentation
- Added tests for Flatten
_ Updated test_tensor

Upabjojr commented:
Commented code needs to be removed.

Upabjojr commented:
There is a conflict with master. You need to merge the master branch and resolve the conflict.

Commit:
Resolved conflict with master

Commit:
Added args test

Commit:
Refactored codes

Commit:
Bug fixed for py2.7

Commit:
Fixed bug with generator

Commit:
Fiexed bug about generator

Upabjojr commented on sympy/core/tests/test_args.py:
```diff
@@ -3982,6 +3982,11 @@ def test_sympy__tensor__array__array_comprehension__ArrayComprehension():
+def test_sympy__tensor__array__arrayop__Flatten():
+    from sympy.tensor.array.arrayop import Flatten
+    from sympy.tensor.array.dense_ndim_array import ImmutableDenseNDimArray
+    fla = Flatten(ImmutableDenseNDimArray(range(24)).reshape(2, 3, 4))
```
where is the call to `_test_args`?

Upabjojr commented on sympy/tensor/array/arrayop.py:
```diff
@@ -296,3 +296,73 @@ def permutedims(expr, perm):
+    >>> A = Array(range(6)).reshape(2, 3)
+    >>> Flatten(A)
+    Flatten([[0, 1, 2], [3, 4, 5]])
+    >>> [*Flatten(A)]
```
is this valid in Python 2.7?

Upabjojr commented on sympy/tensor/array/arrayop.py:
```diff
@@ -296,3 +296,73 @@ def permutedims(expr, perm):
+    def __next__(self):
+        from sympy.matrices.matrices import MatrixBase
+
+        try:
```
please remove this `try`. You should define the total number of elements and then count up to them.

Upabjojr commented on sympy/tensor/array/ndim_array.py:
```diff
@@ -140,10 +140,7 @@ def _handle_ndarray_creation_inputs(cls, iterable=None, shape=None, **kwargs):
-            # Construction from another `NDimArray`:
-            elif isinstance(iterable, NDimArray):
-                shape = iterable.shape
-                iterable = list(iterable)
```
why was this removed???

kangzhiq replied:
Because we are trying to cast an `Array` to `list` and then flatten it in the constructor with `flatten()`. Since `flatten()` can now flatten an `Array` properly, I was thinking that we don't need this case any more.

Upabjojr replied:
OK

Commit:
Removed try except and modified tests
